File: challenges/typescript/src/data-model.ts
export interface Library {
  name: string;
  address: string;
  catalog: Catalog;
  userManagement: UserManagement;
}

export interface Catalog {
  booksByIsbn: Record<ISBN, Book>;
  authorsById: Record<AuthorId, Author>;
}

export interface UserManagement {
  librariansByEmail: Record<MemberEmail, Librarian>;
  membersByEmail: Record<MemberEmail, Member>;
}

export type MemberEmail = string;

export interface Member {
  email: MemberEmail;
  encryptedPassword: string;
  isBlocked: boolean;
}

export interface Librarian {
  email: MemberEmail;
  encryptedPassword: string;
}

export type ISBN = string;

export interface Book {
  title: string;
  publicationYear: number;
  isbn: ISBN;
  authorIds: AuthorId[];
  bookItems: BookItem[];
}

export type  AuthorId = string;

export interface BookItem {
  id: string;
  rackId: string;
  purchaseDate: string;
  isLent: boolean;
}

export interface Author {
  id: AuthorId;
  name: string;
  bookIsbns: ISBN[];
}

File: challenges/c#-2/Tests/DataModel.cs
namespace Tests;

record LibraryData(CatalogData Catalog, UserManagementData UserManagement);

record UserManagementData(
    Map<MemberData>    Members,
    Map<LibrarianData> Librarians);

record LibrarianData;

record MemberData(bool IsBlocked);

record UserData;

record CatalogData(
    Map<BookData>   BooksByIsbn,
    Map<AuthorData> AuthorsById);

record BookItemData;

record AuthorData(string Name, string[] BookIsbns);

record BookData(
    string         Isbn,
    string         Title,
    int            PublicationYear,
    BookItemData[] BookItems,
    string[]       AuthorIds);